gh-1031 refactor C Style Guide (#1692)
tarantool/doc: doc/dev_guide/c_style_guide_new.rst

         "more_magic %reg2, %reg3"
         : /* outputs */ : /* inputs */ : /* clobbers */);

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Chapter 16: Conditional Compilation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Wherever possible, don't use preprocessor conditionals (#if, #ifdef) in .c
files; doing so makes code harder to read and logic harder to follow. Instead,
use such conditionals in a header file defining functions for use in those .c
files, providing no-op stub versions in the #else case, and then call those
functions unconditionally from .c files. The compiler will avoid generating
any code for the stub calls, producing identical results, but the logic will
remain easy to follow.

Prefer to compile out entire functions, rather than portions of functions or
portions of expressions. Rather than putting an ifdef in an expression, factor
out part or all of the expression into a separate helper function and apply the
conditional to that function.

If you have a function or variable which may potentially go unused in a
particular configuration, and the compiler would warn about its definition

> going unused, mark the definition as __maybe_unused rather than wrapping it in
> a preprocessor conditional. (However, if a function or variable *always* goes
> unused, delete it.)

Author: Will this be correct?

```suggestion
going unused, do not compile it and use #if for this.
(However, if a function or variable *always* goes unused, delete it.)
```

-------------------------------------------------------------------------------
                                C Style Guide
-------------------------------------------------------------------------------

We use Git for revision control. The latest development is happening in the
default branch (currently ``master``). Our git repository is hosted on GitHub,
and can be checked out with ``git clone git://github.com/tarantool/tarantool.git``
(anonymous read-only access).

If you have any questions about Tarantool internals, please post them on the
`developer discussion list <https://groups.google.com/forum/#!forum/tarantool>`_
or on `StackOverflow <https://stackoverflow.com/questions/tagged/tarantool>`_.
Additionally, some engineers are always present on #tarantool channel on

> irc.freenode.net.

Reviewer: If we don't care about atomicity of the changes, I would delete the google group and irc.freenode.net mentioning. The latter does not even work.

**General guidelines**

The project's coding style is based on the `Linux kernel coding style
<https://www.kernel.org/doc/html/v4.10/process/coding-style.html>`_.

However, we have some additional guidelines, either unique to Tarantool or
deviating from the Kernel guidelines. Below we cite the Linux kernel
coding style noting Tarantool's style features.

We don't cite chapters 10 "Kconfig configuration files", 11 "Data structures",
13 "Printing kernel messages", and 17 "Don't re-invent the kernel macros" since
they are specific to Linux kernel programming environment.

================================================================================
                           Linux kernel coding style
================================================================================

This is a short document describing the preferred coding style for the
linux kernel. Coding style is very personal, and I won't **force** my
views on anybody, but this is what goes for anything that I have to be
able to maintain, and I'd prefer it for most other things too.  Please

> at least consider the points made here.

Reviewer: This paragraph could make a user think he has a choice not to use this guide, which is not true. We exactly **force** our style.

I guess this is a direct cite from the kernel guide, which adds +1 point to the option of rewriting the guide entirely.

Don't put multiple statements on a single line unless you have
something to hide:

.. code-block:: c

  if (condition) do_this;
    do_something_everytime;

Don't put multiple assignments on a single line either. Kernel coding style
is super simple. Avoid tricky expressions.

Outside of comments and documentation, spaces are never
used for indentation, and the above example is deliberately broken.

Get a decent editor and don't leave whitespace at the end of lines.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Chapter 2: Breaking long lines and strings
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Coding style is all about readability and maintainability using commonly
available tools.

The limit on the length of lines is 80 columns and this is a strongly
preferred limit.

.. admonition:: Tarantool Style
    :class: FACT

    As for comments, the same limit of 80 columns is applied.

Statements longer than 80 columns will be broken into sensible chunks, unless
exceeding 80 columns significantly increases readability and does not hide
information. Descendants are always substantially shorter than the parent and
are placed substantially to the right. The same applies to function headers

> with a long argument list. However, never break user-visible strings such as

Reviewer: We allow to break strings.

.. code-block:: c

    int
    function(int x)
    {
      body of function
    }

Heretic people all over the world have claimed that this inconsistency
is ...  well ...  inconsistent, but all right-thinking people know that
(a) K&R are **right** and (b) K&R are right.  Besides, functions are
special anyway (you can't nest them in C).

Note that the closing brace is empty on a line of its own, **except** in
the cases where it is followed by a continuation of the same statement,
ie a ``while`` in a do-statement or an ``else`` in an if-statement, like
this:

.. code-block:: c

    do {
      body of do-loop
    } while (condition);

and

.. code-block:: c

    if (x == y) {
      ..
    } else if (x > y) {
      ...
    } else {
      ....
    }

Rationale: K&R.

Also, note that this brace-placement also minimizes the number of empty
(or almost empty) lines, without any loss of readability.  Thus, as the
supply of new-lines on your screen is not a renewable resource (think
25-line terminal screens here), you have more empty lines to put
comments on.

Do not unnecessarily use braces where a single statement will do.

.. code-block:: c

    if (condition)
      action();

>     and

Reviewer: "and" is not supposed to be a part of the code block.

.. code-block:: none

    if (condition)
      do_this();
    else
      do_that();

This does not apply if only one branch of a conditional statement is a single
statement; in the latter case use braces in both branches:

.. code-block:: c

    if (condition) {
      do_this();
      do_that();
    } else {
      otherwise();
    }

**********************
Chapter 3.1: Spaces
**********************

> Tarantool style for use of spaces depends (mostly) on

Reviewer: Does not look like a direct cite. And is not different from Linux. Should we even change anything here? I am talking about using "Tarantool" word in the kernel style guide copy-paste.

Use one space around (on each side of) most binary and ternary operators,
such as any of these::

    =  +  -  <  >  *  /  %  |  &  ^  <=  >=  ==  !=  ?  :

but no space after unary operators::

    &  *  +  -  ~  !  sizeof  typeof  alignof  __attribute__  defined

no space before the postfix increment & decrement unary operators::

    ++  --

no space after the prefix increment & decrement unary operators::

    ++  --

and no space around the ``.`` and ``->`` structure member operators.

Do not leave trailing whitespace at the ends of lines. Some editors with
``smart`` indentation will insert whitespace at the beginning of new lines as
appropriate, so you can start typing the next line of code right away.
However, some such editors do not remove the whitespace if you end up not
putting a line of code there, such as if you leave a blank line. As a result,
you end up with lines containing trailing whitespace.

Git will warn you about patches that introduce trailing whitespace, and can
optionally strip the trailing whitespace for you; however, if applying a series
of patches, this may make later patches in the series fail by changing their
context lines.

> .. admonition:: Tarantool Style

Reviewer: This note better be moved a little bit above, to the text talking about unary operators (because this is one).

   but if there is a clear reason for why it under certain circumstances
   might be an ``unsigned int`` and under other configurations might be
   ``unsigned long``, then by all means go ahead and use a typedef.

#. when you use sparse to literally create a **new** type for
   type-checking.

#. New types which are identical to standard C99 types, in certain
   exceptional circumstances.

   Although it would only take a short amount of time for the eyes and
   brain to become accustomed to the standard types like ``uint32_t``,
   some people object to their use anyway.

   Therefore, the Linux-specific ``u8/u16/u32/u64`` types and their
   signed equivalents which are identical to standard types are
   permitted -- although they are not mandatory in new code of your
   own.

   When editing existing code which already uses one or the other set
   of types, you should conform to the existing choices in that code.

#. Types safe for use in userspace.

   In certain structures which are visible to userspace, we cannot
   require C99 types and cannot use the ``u32`` form above. Thus, we
   use __u32 and similar types in all structures which are shared

>    with userspace.

Reviewer: Another +1 point to the option of writing our own style guide - this useless information. We are userspace, so this can't be applied to Tarantool. There are many places like this.

different cases, it's OK to have a longer function.

However, if you have a complex function, and you suspect that a
less-than-gifted first-year high-school student might not even
understand what the function is all about, you should adhere to the
maximum limits all the more closely. Use helper functions with
descriptive names (you can ask the compiler to in-line them if you think
it's performance-critical, and it will probably do a better job of it
than you would have done).

Another measure of the function is the number of local variables. They
shouldn't exceed 5-10, or you're doing something wrong. Re-think the
function, and split it into smaller pieces. A human brain can
generally easily keep track of about 7 different things, anything more
and it gets confused. You know you're brilliant, but maybe you'd like
to understand what you did 2 weeks from now.

In source files, separate functions with one blank line. If the function is
exported, the **EXPORT** macro for it should follow immediately after the
closing function brace line.  E.g.:

.. code-block:: c

  int
  system_is_up(void)
  {
    return system_state == SYSTEM_RUNNING;
  }

>   EXPORT_SYMBOL(system_is_up);

Reviewer: Another unnecessary info from Linux guide - we don't have `EXPORT`.

try to explain HOW your code works in a comment: it's much better to
write the code so that the **working** is obvious, and it's a waste of
time to explain badly written code.

Generally, you want your comments to tell WHAT your code does, not HOW.
Also, try to avoid putting comments inside a function body: if the
function is so complex that you need to separately comment parts of it,
you should probably go back to chapter 6 for a while. You can make
small comments to note or warn about something particularly clever (or
ugly), but try to avoid excess. Instead, put the comments at the head
of the function, telling people what it does, and possibly WHY it does
it.

.. admonition:: Tarantool Style
    :class: FACT

    When commenting the Tarantool C API functions, please use Doxygen comment format,
    Javadoc flavor, i.e. `@tag` rather than `\\tag`.
    The main tags in use are ``@param``, ``@retval``, ``@return``, ``@see``,
    ``@note`` and ``@todo``.

    Every function, except perhaps a very short and obvious one, should have a
    comment. A sample function comment may look like below:

    .. code-block:: c

>         /** Write all data to a descriptor.

Reviewer: We don't write text on the same line as the comment opening (`/**`). It should be empty.

         *
         * This function is equivalent to 'write', except it would ensure
         * that all data is written to the file unless a non-ignorable
         * error occurs.
         *
         * @retval 0  Success

>          *

Reviewer: Don't need to put empty lines between doxygen commands. At least if they are short.

         * @retval  1  An error occurred (not EINTR)

>          * /

Reviewer: It is not a valid C comment closure - should be `*/`, not `* /`.

        static int
        write_all(int fd, void \*data, size_t len);

>     It's also important to comment data, whether they are basic types or derived

Reviewer: data -> data types.

    types. To this end, use just one data declaration per line (no commas for
    multiple data declarations). This leaves you room for a small comment on each

>     item, explaining its use.

Reviewer: I couldn't parse the last 2 sentences.

    (let* ((anchor (c-langelem-pos c-syntactic-element))
           (column (c-langelem-2nd-pos c-syntactic-element))
           (offset (- (1+ column) anchor))
           (steps (floor offset c-basic-offset)))
      (* (max steps 1)
         c-basic-offset)))

  (add-hook 'c-mode-common-hook
            (lambda ()
              ;; Add kernel style
              (c-add-style
               "linux-tabs-only"
               '("linux" (c-offsets-alist
                          (arglist-cont-nonempty
                           c-lineup-gcc-asm-reg
                           c-lineup-arglist-tabs-only))))))

  (add-hook 'c-mode-hook
            (lambda ()
              (let ((filename (buffer-file-name)))
                ;; Enable kernel mode for the appropriate files
                (when (and filename
                           (string-match (expand-file-name "~/src/linux-trees")
                                         filename))
                  (setq indent-tabs-mode t)
                  (setq show-trailing-whitespace t)
                  (c-set-style "linux-tabs-only")))))

This will make emacs go better with the kernel coding style for C

> files below ``~/src/linux-trees``.

Reviewer: This is not related to Tarantool. It assumes you have a special file with rules, and even if you really have it - it contains Linux rules, without our corrections. Another point why citing the kernel guide is a bad idea.

Author: Can I just remove this chapter or the part about `indent` below is ok?

Reviewer: Looks irrelevant as well.

4) forgetting about precedence: macros defining constants using expressions
   must enclose the expression in parentheses. Beware of similar issues with
   macros using parameters.

   .. code-block:: c

     #define CONSTANT 0x4000
     #define CONSTEXP (CONSTANT | 3)

5) namespace collisions when defining local variables in macros resembling
   functions:

   .. code-block:: c

     #define FOO(x)            \
     ({                        \
       typeof(x) ret;          \
       ret = calc_ret(x);      \
       (ret);                  \
     })

   ret is a common name for a local variable - __foo_ret is less likely
   to collide with an existing variable.

   The cpp manual deals with macros exhaustively. The gcc internals manual also
   covers RTL which is used frequently with assembly language in the kernel.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Chapter 11: Allocating memory
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. admonition:: Tarantool Style
    :class: FACT

    Prefer the supplied slab (salloc) and pool (palloc) allocators to malloc()/free()
    for any performance-intensive or large  memory allocations. Repetitive use of

>     malloc()/free() can lead to memory fragmentation and should therefore be avoided.

Reviewer: We don't use `salloc` nor `palloc`. Also `malloc` is not a big deal when is used in a non-perf critical code. For perf critical code there are various options and hacks, not only slab allocator.

Author: It [came](https://www.tarantool.io/en/doc/latest/dev_guide/c_style_guide/#allocating-memory) from the previous version of this guide. Should I remove this statement at all?

Reviewer: I would say something like "prefer specialized allocators like region, mempool, smalloc" instead of "salloc/palloc". For perf critical data, which is almost everything with rare exceptions. "salloc/palloc" didn't ever exist in our code, so the previous version of the guide was also wrong.

technically correct, gcc is capable of inlining these automatically without
help, and the maintenance issue of removing the inline when a second user
appears outweighs the potential value of the hint that tells gcc to do
something it would have done anyway.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Chapter 13: Function return values and names
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Functions can return values of many different kinds, and one of the
most common is a value indicating whether the function succeeded or
failed.  Such a value can be represented as an error-code integer
(-Exxx = failure, 0 = success) or a ``succeeded`` boolean (0 = failure,
non-zero = success).

Mixing up these two sorts of representations is a fertile source of
difficult-to-find bugs.  If the C language included a strong distinction
between integers and booleans then the compiler would find these mistakes
for us... but it doesn't.  To help prevent such bugs, always follow this
convention:

  If the name of a function is an action or an imperative command,
  the function should return an error-code integer.  If the name

>   is a predicate, the function should return a "succeeded" boolean.

Reviewer: Not correct for Tarantool. In 99.99999% of all cases we return 0 on success, non-zero on error (-1 usually). Errors are saved into a diagnostics area which is global per fiber. We never return error codes as a result of a function.

Author: Should I rewrite just the convention according to your comment, or the whole chapter is misleading?

Reviewer: Only the last paragraph in the chapter looks kind of relevant. The rest is misleading.

.. code-block:: c

  asm ("magic %reg1, #42\n\t"
       "more_magic %reg2, %reg3"
       : /* outputs */ : /* inputs */ : /* clobbers */);

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Chapter 16: Conditional Compilation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Wherever possible, don't use preprocessor conditionals (#if, #ifdef) in .c
files; doing so makes code harder to read and logic harder to follow.  Instead,
use such conditionals in a header file defining functions for use in those .c
files, providing no-op stub versions in the #else case, and then call those
functions unconditionally from .c files.  The compiler will avoid generating
any code for the stub calls, producing identical results, but the logic will
remain easy to follow.

Prefer to compile out entire functions, rather than portions of functions or
portions of expressions.  Rather than putting an ifdef in an expression, factor
out part or all of the expression into a separate helper function and apply the
conditional to that function.

If you have a function or variable which may potentially go unused in a
particular configuration, and the compiler would warn about its definition

> going unused, mark the definition as __maybe_unused rather than wrapping it in

Reviewer: We don't have `__maybe_unused`. If a function may not be used, we usually not compile it and use #if for this.

Author: Will this be correct?

```suggestion
going unused, do not compile it and use #if for this.
(However, if a function or variable *always* goes unused, delete it.)
```

Reviewer: Yeah, that is better. We also have C++ templates. I guess it might be a separate task to describe when it is appropriate to use them.

The compiler will constant-fold the conditional away, and include or exclude
the block of code just as with an #ifdef, so this will not add any runtime
overhead.
However, this approach still allows the C compiler to see the code
inside the block, and check it for correctness (syntax, types, symbol
references, etc).  Thus, you still have to use an #ifdef if the code inside the
block references symbols that will not exist if the condition is not met.

At the end of any non-trivial #if or #ifdef block (more than a few lines),
place a comment after the #endif on the same line, noting the conditional
expression used.  For instance:

.. code-block:: c

  #ifdef CONFIG_SOMETHING
  ...
  #endif /* CONFIG_SOMETHING */

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Chapter 17: Header files
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. admonition:: Tarantool Style
    :class: FACT

    Use header guards. Put the header guard in the first line in the header,
    before the copyright or declarations. Use all-uppercase name for the header
    guard. Derive the header guard name from the file name, and append _INCLUDED
    to get a macro name. For example, core/log_io.h -> CORE_LOG_IO_H_INCLUDED. In

>     ``.c`` (implementation) file, include the respective declaration header before

Reviewer: We use `#pragma once` in the new code instead of macro guards.

Author: I need more details to rewrite this part

Reviewer: It is that simple - just use `#pragma once` in the headers. If we need to add some water, then we can say that as the header guards we refer to this construction:
```C
#ifndef THE_HEADER_IS_INCLUDED
#define THE_HEADER_IS_INCLUDED

// ... the header code ...

#endif // THE_HEADER_IS_INCLUDED
```
It works fine, but the guard name `THE_HEADER_IS_INCLUDED` tends to outdate when the file is moved or renamed. This is especially painful with multiple files having the same name in the project, but different path. For instance, we have 3 `error.h` files, which means for each of them we need to invent a new header guard name, and not forget to update them if the files are moved or renamed.

For that reason we use `#pragma once` in all the new code, which shortens the header file down to this:
```C
#pragma once

// ... header code ...
```

(Header guard and pragma once - both prevent inclusion of the same file multiple times in C when `#include "<my_file_name>.h"` is used.)

ask Tarantool developers directly in `telegram <http://telegram.me/tarantool>`_.

**General guidelines**

The project's coding style is inspired by the `Linux kernel coding style
<https://www.kernel.org/doc/html/v4.10/process/coding-style.html>`_.

However, we have some additional guidelines, either unique to Tarantool or
deviating from the Kernel guidelines. Below we rewrite the Linux kernel
coding style according to the Tarantool's style features.

--------------------------------------------------------------------------------
                           Tarantool coding style
--------------------------------------------------------------------------------

This is a short document describing the preferred coding style for the
Tarantool developers and contributors. We **insist** on following these rules
in order to make our code consistent and understandable to any developer.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Chapter 1: Indentation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Tabs are 8 characters, and thus indentations are

> also 8 characters. There are heretic movements that try to make indentations

Reviewer: It is worth mentioning that we use 8-width tabs, not 8 whitespaces.

how the indentation works if you have large indentations.

Now, some people will claim that having 8-character indentations makes
the code move too far to the right, and makes it hard to read on a
80-character terminal screen. The answer to that is that if you need
more than 3 levels of indentation, you're screwed anyway, and should fix
your program.

In short, 8-char indents make things easier to read, and have the added
benefit of warning you when you're nesting your functions too deep.
Heed that warning.

The preferred way to ease multiple indentation levels in a switch statement is
to align the ``switch`` and its subordinate ``case`` labels in the same column
instead of ``double-indenting`` the ``case`` labels. E.g.:

..  code-block:: c

    switch (suffix) {
    case 'G':
    case 'g':
      mem <<= 30;
      break;
    case 'M':
    case 'm':
      mem <<= 20;
      break;
    case 'K':
    case 'k':
      mem <<= 10;
      /* fall through */
    default:
      break;
    }

Don't put multiple statements on a single line unless you have
something to hide:

..  code-block:: c

    if (condition) do_this;
      do_something_everytime;

> Don't put multiple assignments on a single line either. Kernel coding style

Reviewer: It is not Kernel coding style anymore. We must purge all mentioning of the kernel from the doc.

variable that is used to hold a temporary value.

If you are afraid to mix up your local variable names, you have another
problem, which is called the function-growth-hormone-imbalance syndrome.
See chapter 6 (Functions).

For function naming we have a convention is to use:

*    ``new``/``delete`` for functions which
     allocate + initialize and destroy + deallocate an object,
*    ``create``/``destroy`` for functions which initialize/destroy an object
     but do not handle memory management,
*    ``init``/``free`` for functions which initialize/destroy libraries and subsystems.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Chapter 5: Typedefs
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Please don't use things like ``vps_t``.
It's a **mistake** to use typedef for structures and pointers. When you see a

..  code-block:: c

    vps_t a;

in the source, what does it mean?
In contrast, if it says

..  code-block:: c

    struct virtual_container *a;

you can actually tell what ``a`` is.

Lots of people think that typedefs ``help readability``. Not so. They are
useful only for:

> #.  totally opaque objects (where the typedef is actively used to **hide**

Reviewer: `Totally` should start with a capital letter?

    what the object is).

    Example: ``pte_t`` etc. opaque objects that you can only access using
    the proper accessor functions.

    ..  note::

        Opaqueness and ``accessor functions`` are not good in themselves.
        The reason we have them for things like pte_t etc. is that there
        really is absolutely **zero** portably accessible information there.

#.  Clear integer types, where the abstraction **helps** avoid confusion
    whether it is ``int`` or ``long``.

    u8/u16/u32 are perfectly fine typedefs, although they fit into
    point 4 better than here.

    ..  note::

        Again - there needs to be a **reason** for this. If something is
        ``unsigned long``, then there's no reason to do
        typedef unsigned long myflags_t;

    but if there is a clear reason for why it under certain circumstances
    might be an ``unsigned int`` and under other configurations might be
    ``unsigned long``, then by all means go ahead and use a typedef.

> #.  when you use sparse to literally create a **new** type for

Reviewer: Ditto.

    type-checking.

#.  New types which are identical to standard C99 types, in certain
    exceptional circumstances.

    Although it would only take a short amount of time for the eyes and
    brain to become accustomed to the standard types like ``uint32_t``,
    some people object to their use anyway.

>     Therefore, the Linux-specific ``u8/u16/u32/u64`` types and their

Reviewer: We don't use these types. They can only be found in old SQL code, but we are getting rid of them.

complexity and indentation level of that function. So, if you have a
conceptually simple function that is just one long (but simple)
case-statement, where you have to do lots of small things for a lot of
different cases, it's OK to have a longer function.

However, if you have a complex function, and you suspect that a
less-than-gifted first-year high-school student might not even
understand what the function is all about, you should adhere to the
maximum limits all the more closely. Use helper functions with
descriptive names (you can ask the compiler to in-line them if you think
it's performance-critical, and it will probably do a better job of it
than you would have done).

Another measure of the function is the number of local variables. They
shouldn't exceed 5-10, or you're doing something wrong. Re-think the
function, and split it into smaller pieces. A human brain can
generally easily keep track of about 7 different things, anything more
and it gets confused. You know you're brilliant, but maybe you'd like
to understand what you did 2 weeks from now.

In function prototypes, include parameter names with their data types.

> Although this is not required by the C language, it is preferred in Linux

Reviewer: "in Linux" -> "in Tarantool". Might be good to find all the usages of word "Linux" and replace with "Tarantool" where appropriate.

time to explain badly written code.

Generally, you want your comments to tell WHAT your code does, not HOW.
Also, try to avoid putting comments inside a function body: if the
function is so complex that you need to separately comment parts of it,
you should probably go back to chapter 6 for a while. You can make
small comments to note or warn about something particularly clever (or
ugly), but try to avoid excess. Instead, put the comments at the head
of the function, telling people what it does, and possibly WHY it does
it.

When commenting the Tarantool C API functions, please use Doxygen comment format,
Javadoc flavor, i.e. `@tag` rather than `\\tag`.
The main tags in use are ``@param``, ``@retval``, ``@return``, ``@see``,
``@note`` and ``@todo``.

Every function, except perhaps a very short and obvious one, should have a
comment. A sample function comment may look like below:

..  code-block:: c

    /**
     * Write all data to a descriptor.
     *
     * This function is equivalent to 'write', except it would ensure
     * that all data is written to the file unless a non-ignorable
     * error occurs.
     *
     * @retval 0  Success

>      * @retval  1  An error occurred (not EINTR)

Reviewer: Excess whitespaces around `1`.

     */
    static int

>     write_all(int fd, void \*data, size_t len);

Reviewer: Should not have `\` before `*`.

    ..  code-block:: c

        #define FOO(val) bar(index, val)

    might look like a good thing, but it's confusing as hell when one reads the
    code and it's prone to breakage from seemingly innocent changes.

3)  macros with arguments that are used as l-values: ``FOO(x) = y;`` will
    bite you if somebody e.g. turns FOO into an inline function.

4)  forgetting about precedence: macros defining constants using expressions
    must enclose the expression in parentheses. Beware of similar issues with
    macros using parameters.

    ..  code-block:: c

        #define CONSTANT 0x4000
        #define CONSTEXP (CONSTANT | 3)

5)  namespace collisions when defining local variables in macros resembling
    functions:

    ..  code-block:: c

        #define FOO(x)            \
        ({                        \
          typeof(x) ret;          \
          ret = calc_ret(x);      \
          (ret);                  \
        })

    ret is a common name for a local variable - ``__foo_ret`` is less likely
    to collide with an existing variable.

    The cpp manual deals with macros exhaustively. The gcc internals manual also

>     covers RTL which is used frequently with assembly language in the kernel.

Reviewer: Should not mention the kernel probably.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Chapter 11: The inline disease
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

There appears to be a common misperception that gcc has a magic "make me
faster" speedup option called ``inline``. While the use of inlines can be
appropriate, it very often is not. Abundant use of the inline keyword leads to
a much bigger kernel, which in turn slows the system as a whole down, due to a
bigger icache footprint for the CPU and simply because there is less memory
available for the pagecache. Just think about it; a pagecache miss causes a
disk seek, which easily takes 5 milliseconds. There are a LOT of cpu cycles
that can go into these 5 milliseconds.

A reasonable rule of thumb is to not put inline at functions that have more
than 3 lines of code in them. An exception to this rule are the cases where
a parameter is known to be a compiletime constant, and as a result of this
constantness you *know* the compiler will be able to optimize most of your
function away at compile time. For a good example of this later case, see

> the kmalloc() inline function.

Reviewer: `kmalloc()` is a kernel function, we should not mention it.

technically correct, gcc is capable of inlining these automatically without
help, and the maintenance issue of removing the inline when a second user
appears outweighs the potential value of the hint that tells gcc to do
something it would have done anyway.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Chapter 12: Function return values and names
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Functions can return values of many different kinds, and one of the
most common is a value indicating whether the function succeeded or
failed.

In 99.99999% of all cases in Tarantool we return 0 on success, non-zero on error
(-1 usually). Errors are saved into a diagnostics area which is global per fiber.
We never return error codes as a result of a function.

Functions whose return value is the actual result of a computation, rather
than an indication of whether the computation succeeded, are not subject to
this rule. Generally they indicate failure by returning some out-of-range
result. Typical examples would be functions that return pointers; they use

> NULL or the ERR_PTR mechanism to report failure.

Reviewer: We don't have `ERR_PTR`, only `NULL`.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Chapter 13: Editor modelines and other cruft
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Some editors can interpret configuration information embedded in source files,
indicated with special markers. For example, emacs interprets lines marked
like this:

..  code-block:: c

    -*- mode: c -*-

Or like this:

..  code-block:: c

    /*
    Local Variables:
    compile-command: "gcc -DMAGIC_DEBUG_FLAG foo.c"
    End:
    */

Vim interprets markers that look like this:

..  code-block:: c

    /* vim:set sw=8 noet */

Do not include any of these in source files. People have their own personal
editor configurations, and your source files should not override them. This
includes markers for indentation and mode configuration. People may use their
own custom mode, or may have some other magic method for making indentation
work correctly.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Chapter 14: Inline assembly

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: Tbh, I would drop chapter this entirely. We don't write in assembly except extremely rare exceptions which I can't even remember what was the last one.

string, and end each string except the last with ``\n\t`` to properly indent the
next instruction in the assembly output:

..  code-block:: c

    asm ("magic %reg1, #42\n\t"
         "more_magic %reg2, %reg3"
         : /* outputs */ : /* inputs */ : /* clobbers */);

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Chapter 15: Conditional Compilation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Wherever possible, don't use preprocessor conditionals (``#if``, ``#ifdef``) in
.c files; doing so makes code harder to read and logic harder to follow. Instead,
use such conditionals in a header file defining functions for use in those .c
files, providing no-op stub versions in the #else case, and then call those
functions unconditionally from .c files. The compiler will avoid generating
any code for the stub calls, producing identical results, but the logic will
remain easy to follow.

Prefer to compile out entire functions, rather than portions of functions or
portions of expressions. Rather than putting an ``#ifdef`` in an expression,
factor out part or all of the expression into a separate helper function and

> apply the conditional to that function.

Reviewer: Shouldn't it be "the conditional" -> "the condition"?

If you have a function or variable which may potentially go unused in a
particular configuration, and the compiler would warn about its definition
going unused, do not compile it and use #if for this.
(However, if a function or variable *always* goes unused, delete it.)

> Within code, where possible, use the IS_ENABLED macro to convert a Kconfig

Reviewer: We don't have `IS_ENABLED` nor `Kconfig`. I would propose to drop the text from here to line 841.
